### ISSUE TITLE
Python 3.10 manylinux wheels

### ISSUE THREAD

**Issue by hrnciar** [numpy/numpy#19630]
Hello.

Could you please release CPython 3.10 manylinux wheels on PyPI? 

Python 3.10.0 release candidate 1 was released this week: https://www.python.org/downloads/release/python-3100rc1/

It should be quite safe to release wheels for release candidate. ABI for Python 3.10 should be frozen from now on, even for bugfixes. From release notes: 

> There will be no ABI changes from this point forward in the 3.10 series and the goal is that there will be as few code changes as possible.

Thank You.

**Comment by tylerjereddy** [comment]
Matching issue in SciPy is cross-linked above--I'll keep an eye on what Chuck/the NumPy team decides to do.

**Comment by charris** [comment]
To build wheels for 3.10 currently requires that it be available in manylinux2014 (linux), and azure (windows and mac). Last time it was about three weeks after 1.9 was released before we could build the wheels.

**Comment by charris** [comment]
Note that rc1 did break a couple of NumPy tests.

**Comment by charris** [comment]
Checking manylinux2014, likely the first out the door, they currently support 3.10.0b2. I imagine that will be updated in the next few weeks, maybe when rc2 comes out.

**Comment by charris** [comment]
Hmm, it does look like manylinux2014 does support rc1.

> 3.10.0rc1 is now in the latest images.
ABI is now stable: https://www.python.org/downloads/release/python-3100rc1/

There was no PR for that, it must have just been pushed. I think we can release linux wheels for that.

**Comment by charris** [comment]
I can build manylinux2014 Python3.10.0rc1 wheels, but they can't be tested until 3.10 is added to the testing docker images. @matthew-brett @mattip Who does that?

**Comment by mattip** [comment]
Me. The repo is https://github.com/multi-build/docker-images, I will update it.

**Comment by mattip** [comment]
I updated the docker images. The version available via `apt get` after adding the deadsnake apt repo is 
```
Python 3.10.0a6 (default, Mar  2 2021, 02:01:08) [GCC 5.4.0 20160609] on linux
```

I hope that is recent enough? Note these are xenial images, but xenial is EOL. I guess I should update the whole thing to focal (20.04)

**Comment by hrnciar** [comment]
> I updated the docker images. The version available via `apt get` after adding the deadsnake apt repo is
> 
> ```
> Python 3.10.0a6 (default, Mar  2 2021, 02:01:08) [GCC 5.4.0 20160609] on linux
> ```
> 
> I hope that is recent enough? Note these are xenial images, but xenial is EOL. I guess I should update the whole thing to focal (20.04)

This seems quite outdated, the latest released version is `Python 3.10.0rc1`.

**Comment by mattip** [comment]
The 3.10rc1 is available in the focal images. But now a new problem: ubuntu has dropped 32-bit i686. How much longer will we ship wheels for that platform?

**Comment by hroncok** [comment]
Honestly, I would not bother shipping 32-bit i686 wheels for recent Python versions. Ubuntu has dropped it, so did Fedora...

**Comment by charris** [comment]
@mattip, I'm trying to recall why we use different images for building and testing, something to do with versions?

**Comment by charris** [comment]
New error: `undefined symbol: _PyGen_Send`

**Comment by charris** [comment]
@mattip The 64 bit wheels on Intel now work, but we need `multibuild/focal_arm64v8` for ARM64.

**Comment by charris** [comment]
The Python 3.10 wheels for Intel 64 bit linux should be available from the nightly builds in an hour or so.  You can install them like so:
```
pip install -i https://pypi.anaconda.org/scipy-wheels-nightly/simple numpy
```

**Comment by hroncok** [comment]
I am not sure why this was closed. No 3.10 manylinux wheels are available on PyPI.

**Comment by charris** [comment]
@hroncok Because we aren't making an official release with 3.10 wheels until 3.10 is released. The wheel is for testing, in particular for downstream projects who test against the prereleases.

**Comment by hroncok** [comment]
I don't understand why not. Could you please explain this? I think that having the wheels on PyPI before the final Python 3.10 release (instead of introducing an arbitrary delay after is is released) is beneficial to everybody. Note that:

 1. The ABI is locked now. No ABI changes will happen before the final release. There is no reason to wait.
 2. Even if the above promise ends up broken, wheels have release numbers and can be re-generated for a single version.

**Comment by hroncok** [comment]
Note that the users will not know about pypi.anaconda.org/scipy-wheels-nightly, they will run `pip install numpy` instead.

**Comment by hroncok** [comment]
Clarification: I understand the intention to release the wheels for testing first. However let's not close this issue until the wheels are (tested and) released on PyPI?

**Comment by hroncok** [comment]
> I think that having the wheels on PyPI before the final Python 3.10 release (instead of introducing an arbitrary delay after is is released) is beneficial to everybody.

One more thing that come to my mind after I posted this: Other projects that would like to release their 3.10 wheels [are waiting for this](https://github.com/matplotlib/matplotlib/pull/20819). Waiting after the final release of Python 3.10 only makes the delay chained. The time window between the rc1 and final is exactly the time to figure all this out, so users can enjoy 100% pain-free experience since the Python 3.10 release day.

**Comment by mattip** [comment]
> we need multibuild/focal_arm64v8 for ARM64.

It [is now available](https://hub.docker.com/r/multibuild/focal_arm64v8).

**Comment by brettcannon** [comment]
As  CPython core dev I would say having projects have the ability to get a 3.10 wheel built for testing during the RC at least -- if not the betas -- would be beneficial to both Python itself and the general community as having dependencies that one can't get built locally means everyone ends up postponing supporting new Python releases for months as dependencies trickle out support.

**Comment by mattip** [comment]
@brettcannon could you comment on the policy not to release wheels on PyPI rather to make wheels available through a secondary index, in this case `pip install -i https://pypi.anaconda.org/scipy-wheels-nightly/simple numpy` ? What would the PyPA consider as best practice and when should wheels appear on PyPI for a new python version? How should those wheels be labelled to indicate they may not be final, and what are the appropriate command line arguments for wheel building to generate those labels?

**Comment by brettcannon** [comment]
@mattip no official PyPA position, but https://snarky.ca/what-to-do-when-you-botch-a-release-on-pypi/ outlines what you can do using wheel build numbers.

**Comment by mattip** [comment]
Thanks. So I guess we will have to forge our own path here. @hroncok could you pitch your case on the mailing list for having python pre-release wheels on PyPI and add a suggestion how best to handle a scenario where the pre-release wheels are broken by a newer release candidate or final release? Is it via build numbers or should we add some qualifier to the pre-release wheels and then always add a post-release wheel update?

**Comment by encukou** [comment]
The secondary index is problematic because the build mechanisms of dependent projects need to be modified, to pass flags to `pip` depending on the Python version.

CPython's policy for changes between `3.10.0rc1` and `3.10.0 final` is *stricter* than for changes between `3.9.5` and `3.9.6`. (Not much, but stricter – only reviewed bug fixes are allowed.)
How would you to handle the scenario where `3.9.6` breaks published wheels? Handle breakage in `3.10.0 final` the same way. Probably that'd be an emergency bugfix release of CPython?

I'm not sure PyPA can easily speak as a group. But maybe we can get @pablogsal to add a statement to the release announcement?

As I personally see it, now it's time for *integration* and *integration testing* in the ecosystem – finding the bugs that happen when everything is hooked up. Ideally the 3.10.0 final release should just be a removal of the `rc` tag, not a code change.

**Comment by hroncok** [comment]
Something like this? What mailing list do you want me to share this at, numpy-discussion?

## Why publish Python pre-release wheels of NumPy, SciPy, etc. on PyPI?

### Reason 1: The wheels should be available on PyPI at the time of the Python 3.10 final release

As Python 3.10 final comes out in October, many users will start migrating to it, either by their own choice at their own schedule, or by choice of their system/package manager distributor. For example, the Python that is installed via the app store in Windows will be updated to 3.10, Python in homebrew will be updated to 3.10, users of Fedora 35 (that is to be released in October as well) will have Python 3.10 b default. Many users will use Python 3.10 since day 0. Most of those users will use pip with the default package index: PyPI.

I think that to provide a **smooth user experience**, important projects on PyPI with extension modules (such as NumPy or SciPy, MarkupSafe, Matplotlib, Pandas...) need to have binary wheels available when Python 3.10.0 is released, not later. When the wheels are delayed, the user experience will be significantly worse, reminding the situation from a couple of years back, when you needed to compile NumPy to install it with pip and users, especially beginners, were contused by the failures. Tribal knowledge was required to fetch some projects from alternate locations or even via alternate installers. Now, the situation would be even worse, because the tribal knowledge would need to differ for each Python version.

**By publishing the wheels on PyPI since rc1, we get time to achieve this.** If only we had a magical thing that would publish all the wheels on PyPI the second Python 3.10.0 is released, but we don't. Adapting the CI to produce the wheels, coordinating with the dependencies, verification, it all takes time, which brings me to... 

###  Reason 2: Dependent projects use packages from PyPI (on the CI and locally)

Since publishing the wheels is more or less a chain-action (SciPy is blocked on NumPy, something else is blocked on SciPy...), it may take several weeks (or months) to publish the wheels.

Technically, SciPy can build the wheel using NumPy compiled from sources, but it breaks and needs more resources to figure it out and special case it. Hence it is easier to wait for NumPy.

Similarly, the CI of SciPy could special case the location of the NumPy wheels to continue . But there are hundreds (thousands?) of projects that depend on NumPy and they want to run Python 3.10 on their CI even with pre-releases - both to test their code with it or to also produce wheels. They want to run stuff like `tox -e py310` on their CI, not `if python_version == 3.10: download numpy from X`.

Well, they even want to run `tox -e py310` locally and they don't want to compile NumPy and SciPy every time they recreate the virtual environment.

**By publishing the wheels on PyPI since pre-releases, we get more projects tested with the next Python release.** And we discover more bugs in Python as well.

### Challenge 1: What if the wheels are no longer ABI compatible with the next pre-release/final release?

You seem to be concerned that if you publish a wheel built with Python 3.10.0rc1, it might be broken with Python 3.10.0rc2 or Python 3.10.0 final. Others have already covered this topic here, so let's sum it up:

 - **ABI compatibility guarantees** between Python 3.10.0rc1, Python 3.10.0rc2 and Python 3.10.0 final are **stronger** than between 3.9.6 and 3.9.7. You don't seem to worry about releasing wheels built with Python 3.9.6.
 - There is [a mechanism to publish a newly re-built wheel](https://snarky.ca/what-to-do-when-you-botch-a-release-on-pypi/) of the same release/code on PyPI.

### Challenge 2: What if the code is no longer compatible with the next pre-release/final release?

Not every compatibility is ABI compatibility. What if NumPy releases a wheel that was tested on Python 3.10.0rc1 but there is an incompatibility with the final release that needs to be fixed in NumPy's code? (Note that again, I consider this extremely unlikely and similar or smaller risk than with every Python bugfix release.)

Well, you would cut a new release. There is no difference in (not) publishing the wheels on PyPI: the incompatibility will also hit users who actually installed NumPy from source, so you still need to do this, even if you don't publish the wheels.

### Challenge 3: What if users install NumPy on pre-releases of Python we don't intend to fully support yet?

The users can already do that even without wheels. By not publishing the wheels, it is more likely to break, but it will not stop them nor signalize the support level.

**Comment by mattip** [comment]
@hroncok thanks, well stated and it convinced me :). In fact, this is so well written that maybe it should become a blog post or added to the PyPA docs somewhere. I will wait for someone else from NumPy to chime in with other thoughts about whether we need to ask the numpy-discussions mailing list.

We are still using Azure pipelines for wheel building on windows, xref MacPython/numpy-wheels#116 to move to github actions

**Comment by charris** [comment]
We could  make patch releases with 3.10 wheels as long as it is clear that they are experimental. NumPy 1.21.2 will come out within a week and we could try doing that. It is planned that NumPy 1.21 will support 3.10 in any case, the question is timing and platform coverage.

**Comment by rgommers** [comment]
While I'm perfectly happy with uploading 3.10 wheels for NumPy and other packages, I do want to point out a few things:
- We are already asking maintainers of packages that depend on NumPy to use https://pypi.anaconda.org/scipy-wheels-nightly/simple/ in their CI to test against NumPy master. And scipy/pandas/etc. wheels are there as well. So it's not that strange a location.
- @charris had already anticipated supporting Python 3.10 in Numpy 1.21.x and we had CI in place, therefore the `python_requires` had a `<3.11.0` upper bound. But in general this will not always be true. SciPy 1.7.x has `<3.10.0`, so it is not correct to upload wheels to the current latest release - that will require a new `1.7.2` or `1.8.0` release. (and yes, the metadata is right - all dependencies should have a correct upper bound)

> For example, the Python that is installed via the app store in Windows will be updated to 3.10, Python in homebrew will be updated to 3.10, users of Fedora 35 (that is to be released in October as well) will have Python 3.10 by default. Many users will use Python 3.10 since day 0.

This is not a good thing, those distributors are causing problems by doing that on day zero. Python is a little bit special, but not *that* special. It is just one dependency in a very large dependency graph, so this makes very little sense. This super-agressive upgrading should be discouraged. Because the end result will indeed be "beginners end up being confused".

Any project normally has a choice between:
1. guaranteeing API and ABI stability, so a downstream project building against release `x.y` will have packages up that also work with release `x.y+1` (this is what NumPy does).
2. retaining the freedom to break API or ABI, at which point you pay with delayed uptake and "early adopter" issues.

Python is doing the first half of (2) and is asking for extra work/releases from their downstream packages to avoid paying the price of (2). If it's the only project that does that (and I think that's the case), then that's still a cost that can be paid fairly easily, however if other packages who choose to not guarantee ABI stability (PyTorch and TensorFlow come to mind) would do the same, that cost quickly gets out of hand. It's not a great habit.

**Comment by hroncok** [comment]
The rc versions of Python do guarantee ABI stability, which was exactly the point why the wheels were requested after rc went out.

**Comment by rgommers** [comment]
I understand that, but that's only a 2 month period. It's a very nice improvement, don't get me wrong. However, not all projects are going to do a release (at all, or with Python 3.10 support) in those two months. So it is still going to cause problems if distributors of Python cause beginners to end up with 3.10 straight away on release.

**Comment by encukou** [comment]
> I understand that, but that's only a 2 month period.

What other time period would you suggest to use? Your arguments could just as well apply to two years after the freeze.

**Comment by rgommers** [comment]
Eh no, they don't. Most important Python projects are using a 3-9 month release cycle, with the typical spacing for minor releases being 6 months.

@encukou I think you're missing the main point. I do not expect CPython to suddenly become ABI stable between minor releases, and a 2 month window is already a lot nicer than nothing. The problem is the expectation that it's okay to have tons of users upgrade to a new release and expect it to all work fine. What Fedora, Homebrew and the Windows App Store are doing is the main problem here, especially coupled with the expectation that it's a good idea to use all those Python install methods together with PyPI.

**Comment by encukou** [comment]
Ah, so the issue is in *upgrading*? Yes, I guess all Python distributors should either preserve the old version of Python on upgrade, or not make the upgrade automatic.
I thought we're talking more about *new installs* of Python – if we agree that new users should get 3.10 by default when 3.10 comes out, then we do need wheels for 3.10 soon.

> Most important Python projects are using a 3-9 month release cycle

Important projects are also tested with alphas/betas. (Not necessarily by the upstream maintainers.) For rc1 they can either "just" build new wheels, or (in hopefully rare cases) make a point release with a specific fix.

> I do not expect CPython to suddenly become ABI stable between minor releases

Not suddenly, but there is a limited API subset that [is ABI stable](https://docs.python.org/3.10/c-api/stable.html#stable), and there's a bunch of effort ongoing to make it more useful. I don't expect NumPy to use it any time soon – as it's be slower than the less stable API – but for some projects the performance/maintenance cost tradeoff should be worth it.

**Comment by rgommers** [comment]
> I thought we're talking more about _new installs_ of Python – if we agree that new users should get 3.10 by default when 3.10 comes out, then we do need wheels for 3.10 soon.

We are talking about both upgrades and new installs. New users should not get 3.10 by default on day 1. Unless they understand the tradeoffs and make a conscious choice that they want to be early adopters. But for something like the Windows App Store, that's certainly not the case. The default in Windows App Store should be 3.9.x for some time (1-3 months?) after 3.10 release date. 

There just aren't enough new/improved things in Python 3.10 that that outweighs some libraries not being installable at all and beginners wasting a lot of time.

> For rc1 they can either "just" build new wheels

I'm sure you are well aware of how complex it is to build wheels in general, and that's why you're using quotes here. It takes a lot of work - we have to move the NumPy wheel build machinery from Azure DevOps to GitHub Actions it looks like for example. Projects more complex than NumPy may have other blockers: maintainer bandwidth, QA trajectories for any official binaries, size limit issues on PyPI that take weeks to get a response, etc.

I suggest to make an inventory after the release of what is missing. My prediction is that despite requests like this issue and people spending time on it, there will be a lot of libraries especially for scientific computing and ML/AI, that will not be available for 3.10 by the release date.

**Comment by encukou** [comment]
> The default in Windows App Store should be 3.9.x for some time (1-3 months?) after 3.10 release date.

A Windows question would be for @brettcannon, I believe?

Can/should we determine who new users are and separate them from old users who are better early adopters? (And should we rather use "early adopter" for users of things like Fedora Rawhide prereleases, which had 3.10 since the alphas?)

AFAICS, the current schedule is the RC period (2 months) for the ecosystem to prepare & test, and the final release is for all users. I don't think we can improve that for 3.10. Should we add another release date for App-Store-like releases for 3.11?

> I'm sure you are well aware of how complex it is to build wheels in general, and that's why you're using quotes here.

Yes, exactly.
But there are also projects for which this is easier (or zero-effort if they're pure-python), but they're blocked by the more heavy-weight foundational projects. As I see it, NumPy is in a small category of hard+urgent packages.

> I suggest to make an inventory after the release of what is missing.

Definiely.
The [3.9 wheels issue](https://github.com/numpy/numpy/issues/17482) was based on a user report; this one was filed [proactively](https://bugzilla.redhat.com/show_bug.cgi?id=1989081) based on the "inventory" for 3.9.
If you have any additions to the "inventory", let me know :)

**Comment by brettcannon** [comment]
Nope, a Windows question is a @zooba question.

**Comment by encukou** [comment]
> my current smoke test is "can I `pip install jupyter ...` and render a chart in a notebook"

Thanks! That's essentially the same packages we are looking at. It's nice to have a confirmation :)

**Comment by rgommers** [comment]
> Now, the default `python.exe` stub that ships with Windows gets updated on its own schedule. Right now, that's basically when I feel comfortable with ecosystem status (my current smoke test is "can I `pip install jupyter ...` and render a chart in a notebook"). If the ecosystem is lagging, it won't be switched - certainly doesn't happen automatically on the first day.

That's great to hear, thanks for the explanation @zooba.

> If you have any additions to the "inventory", let me know :)

Beyond what's in https://bugzilla.redhat.com/show_bug.cgi?id=1989081, I'd suggest (packages with compiled code only; pure Python stuff like Dask and Xarray will work already):

- Accelerators: Cython, Numba
- Other scientific computing libraries: CuPy, scikit-learn, scikit-image, statsmodels, spaCy
- Deep learning libraries: PyTorch (I just opened an issue for it, linked above), TensorFlow, and JAX/jaxlib.

**Comment by hroncok** [comment]
Cython has a Pure Python universal wheel available, so the manylinux etc. wheel is not blocking, just faster.
